sdk/core/azure-core/inc/azure/core/internal/client_options.hpp

```diff
@@ -41,29 +41,44 @@ namespace Azure { namespace Core { namespace Internal {
      * @brief Move each policy from \p options into the new instance.
      *
      */
-    ClientOptions(ClientOptions&& options) = default;
+    ClientOptions(ClientOptions&&) = default;
```

Comment:
I've been thinking about this class and I have some ideas. The purpose of this base class is to allow Core to add policies and customer configuation over time without making changes to client library code. Tis is great for many policys like Retry, Transport, & Logging.

But, some policies REQUIRE client library intervention. For example, telemetry requires that the client library apss its name and version and the api-version requires that the client library pass the query parameter/header name and value. For policies that REQUIRE client library intervention, option fields should NOT be in ClientOption and should be in the client library's class that is derived from ClientOptions. For telemetry, for example, the ClientOptions derived class would add a field for ApplicationId and then in the XxxClient ctor, the client libary would:

- Grab this field value and add the client library-specific values necessary to create the TelemetryOptions 
- Pass the TelemetryOptions to the TelemetryPolicy ctor
- Add the policy to the ClientOptions's per-operation list of policies 
- Pass the ClientOptions to the Pipeline Builder

The pipeline builder would NOT construct the TelemetryPolicy or any policy that sets the api-version (if we decide to solve this problem via a policy [more discussion to be had on this]).

This design makes it very clear whose responsibilty everything is. The builder has EVERYTHING it needs to construct the pipeline. And Core can add new policies for customers and all client libraries will automatically jsut supprot them. Any policies that REQUIRE client library intervention is NOT in ClientOptions and is NOT in the pipeline builder - it is the ENTIRELY the responsibility of the client library. 

We should minimize the number of policies that require client library intervention.
Also, one problem with the above design is that we can't change a policy that doesn't require client library intervention to require client library intervention in the future as this would remove it from ClientOptions and the builder which is a breaking change. But, I don't forsee this happenning at all in the future. Should we ever need to cross this bridge (which I seriously doubt), then I have some ideas as to what we could possibly do.

Reply:
so, basically, if the policy requires service-information, it is up to the service to decide if they want to add or not the policy?

Reply:
@JeffreyRichter  This is a very good suggestion to core sdk. I won't handle it in this PR. maybe @vhvb1989 will handle it in a separate PR.